cpp/src/utilities/column_utils.cpp

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2019 BlazingDB, Inc.
+ *     Copyright 2019 Eyal Rozenberg <eyalroz@blazingdb.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "column_utils.hpp"
+#include <utilities/error_utils.hpp>
+
+// This should go up into more general utility code...
+static constexpr inline bool logical_xor(bool x, bool y) { return (x and (not y)) or ((not x) and y) ; }
+
+namespace cudf {
+
+/**
+ * @brief Ensures a gdf_column is valid, i.e. that its fields are consistent
+ * with each other, and logical in themselves, in representing a proper column.
+ */
+void validate(const gdf_column& column)
+{
+    CUDF_EXPECTS ( (column.data != nullptr or column.size == 0)          , "No column data despite there supposedly being rows.");
+    CUDF_EXPECTS ( (column.dtype != GDF_invalid)                         , "Cannot use the 'invalid' column element type.");
+    CUDF_EXPECTS ( (column.dtype < N_GDF_TYPES)                          , "Unknown column data type");
+    CUDF_EXPECTS ( (is_nullable(column) or (column.null_count == 0))     , "Column supposedly has null values, but no place to expose its record");
+    CUDF_EXPECTS ( (column.null_count <= column.size)                    , "Column claims to have more null elements than elements overall");
+}
+
+void validate(const gdf_column* column_ptr)
+{
+    CUDF_EXPECTS ( (column_ptr != nullptr) , "Null gdf_column pointer");
+    validate(*column_ptr);
+}
+
+bool have_matching_types(const gdf_column& validated_column_1, const gdf_column& validated_column_2) noexcept
```

Comment:
Why doesn't this throw exceptions?
I often check input and output columns, usually dtype, size, null check is required, but somtimes dtype can be different.
How about adding this?

```
typedef enum
{
  DTYPE_CHECK = 0x000000001,
  SIZE_CHECK  = 0x000000002,
  NULL_CHECK  = 0x000000004,
  FULL_CHECK  = DTYPE_CHECK | SIZE_CHECK | NULL_CHECK,
} gdf_validate_flag;


void validate(const gdf_column& validated_column_1, const gdf_column& validated_column_2, gdf_validate_flag checkflag=FULL_CHECK)
{
  if( checkflag | DTYPE_CHECK )CUDF_EXPECTS ( (validated_column_1.dtype == validated_column_2.dtype) , "2 columns have different dtypes");
  if( checkflag | SIZE_CHECK )CUDF_EXPECTS ( (validated_column_1.size == validated_column_2.size) , "2 columns have different sizes");
  if( checkflag | NULL_CHECK )CUDF_EXPECTS ( (is_nullable(validated_column_1) == is_nullable(validated_column_2)) , "2 columns have different nulls");
}
```

Reply:
Do you mean `have_matching_types()`? It doesn't throw because it returns a boolean. It's a predicate.

As for vaidate() - I validate everything. Is it important to only make some of the checks?